tools/merger.py

```diff
@@ -72,78 +72,68 @@ def load_merger(filenames, return_mass=False, trim=True):
+    
+    # Fill the data into the empty tables
+    j = 0
+    jp = 0
+    for i, fn in enumerate(filenames):
+        print(f"File number {i+1:d} of {len(filenames)}")
         f = asdf.open(fn, lazy_load=True, copy_arrays=True)
         fdata = f['data']
-
-        N_halos = fdata[fields[0]].shape[0]
-        num_progs_cum = 0
-        if i == 0:
-
-            dtypes = []
-            for field in fields:
-                dtype = fdata[field].dtype
-                try:
-                    shape = fdata[field].shape[1]
-                    dtype = (dtype, shape)
-                except:
-                    pass
-                if field != "Progenitors":
-                    dtypes.append((field, dtype))
-
-                if field == "Progenitors":
-                    final_progs = fdata["Progenitors"]
-                    try:
-                        num_progs = fdata["NumProgenitors"] + num_progs_cum
-                    except:
-                        print(
-                            "You need to also request 'NumProgenitors' if requesting the 'Progenitors' field"
-                        )
-                        exit()
-                    num_progs_cum += np.sum(num_progs)
-
-            final = np.empty(N_halos, dtype=dtypes)
-            for field in fields:
-                if "Progenitors" != field:
-                    final[field] = fdata[field]
-
-            if "Progenitors" in fields:
-                final["NumProgenitors"] = num_progs
-
-        else:
-            new = np.empty(N_halos, dtype=dtypes)
-            for field in fields:
-
-                if field != "Progenitors":
-                    new[field] = fdata[field]
-
-                if field == "Progenitors":
-                    progs = fdata["Progenitors"]
-                    final_progs = np.hstack((final_progs, progs))
-
-                    num_progs = fdata["NumProgenitors"] + num_progs_cum
-                    num_progs_cum += np.sum(num_progs)
-
-            if "Progenitors" in fields:
-                new["NumProgenitors"] = num_progs
-
-            final = np.hstack((final, new))
-        del fdata
-        f.close()
-
-    #cols = {col:np.empty(N_halos, dtype=user_dt[col]) for col in fields}
-    # TESTING
-    final = Table(final, copy=False)
-    if "Progenitors" in fields:
-        return final, final_progs
-
-    return final
+        thisN = len(fdata[fields[0]])
+        
+        for field in fields:
+            # Insert the data into the next slot in the table
+            t[field][j:j+thisN] = fdata[field]
+            
+        if do_prog:
+            thisNp = len(fdata['Progenitors'])
+            p['Progenitors'][jp:jp+thisNp] = fdata['Progenitors']
+            jp += thisNp
+            
+        j += thisN
+    
+    # Should have filled the whole table!
+    assert j == Ntot
+    
+    ret = dict(merger=t)
+    if do_prog:
+        ret['progenitors'] = p
+        assert jp == N_prog_tot
+    
+    return ret
```

Comment:
Now this returns a dict, rather than a variable number of tuple elements.  So upstream code only needs to know one call signature.